File: src/utils/formatting.js
/**
 * Utility functions for formatting numbers, currency, dates, and labels.
 */

// Format currency with support for Indian Rupee (Lakhs/Crores) or USD (K/M)
export function formatCurrency(value, currency = 'INR') {
  if (value === null || value === undefined || isNaN(Number(value))) return '-';
  const num = Number(value);

  if (currency === 'INR') {
    const abs = Math.abs(num);
    const sign = num < 0 ? '-' : '';
    if (abs >= 10000000) {
      return `${sign}₹${(abs / 10000000).toFixed(2)}Cr`;
    }
    if (abs >= 100000) {
      return `${sign}₹${(abs / 100000).toFixed(2)}L`;
    }
    if (abs >= 1000) {
      return `${sign}₹${(abs / 1000).toFixed(1)}k`;
    }
    return `${sign}₹${abs.toLocaleString('en-IN')}`;
  }

  // Fallback to international currency formatting
  return new Intl.NumberFormat('en-US', {
    style: 'currency',
    currency: currency,
    maximumFractionDigits: 2,
    notation: num >= 1000000 ? 'compact' : 'standard',
  }).format(num);
}

// Format numbers compactly
export function formatCompactNumber(value) {
  if (value === null || value === undefined || isNaN(Number(value))) return '-';
  const num = Number(value);
  const abs = Math.abs(num);
  const sign = num < 0 ? '-' : '';

  if (abs >= 10000000) {
    return `${sign}${(abs / 10000000).toFixed(2)}Cr`;
  }
  if (abs >= 100000) {
    return `${sign}${(abs / 100000).toFixed(1)}L`;
  }
  if (abs >= 1000000) {
    return `${sign}${(abs / 1000000).toFixed(2)}M`;
  }
  if (abs >= 1000) {
    return `${sign}${(abs / 1000).toFixed(1)}K`;
  }
  return num.toLocaleString();
}

// Format percentage: 0.184 -> 18.4% or 18.4 -> 18.4%
export function formatPercentage(value) {
  if (value === null || value === undefined || isNaN(Number(value))) return '-';
  let num = Number(value);
  // If provided as a fraction like 0.184, convert to percentage
  if (Math.abs(num) <= 1 && num !== 0) {
    num = num * 100;
  }
  const sign = num > 0 ? '+' : '';
  return `${sign}${num.toFixed(1)}%`;
}

// Format readable labels from snake_case or camelCase keys
export function formatLabel(key) {
  if (!key) return '';
  return key
    .replace(/_/g, ' ')
    .replace(/([A-Z])/g, ' $1')
    .replace(/\b\w/g, (char) => char.toUpperCase())
    .trim();
}

// Smart auto-formatter for table cells based on column key and value
export function formatSmartCell(key, value) {
  if (value === null || value === undefined) return '-';
  const lowerKey = (key || '').toLowerCase();

  // Boolean
  if (typeof value === 'boolean') {
    return value ? 'Yes' : 'No';
  }

  // Check if string date
  if (typeof value === 'string' && /^\d{4}-\d{2}-\d{2}/.test(value)) {
    try {
      const date = new Date(value);
      if (!isNaN(date.getTime())) {
        return date.toLocaleDateString(undefined, { year: 'numeric', month: 'short', day: 'numeric' });
      }
    } catch {
      return value;
    }
  }

  // Numeric checks
  if (typeof value === 'number' || (!isNaN(Number(value)) && typeof value === 'string' && value.trim() !== '')) {
    const num = Number(value);

    // Revenue / Amount / Sales / Cost
    if (lowerKey.includes('revenue') || lowerKey.includes('sales') || lowerKey.includes('amount') || lowerKey.includes('price') || lowerKey.includes('cost') || lowerKey.includes('budget')) {
      return formatCurrency(num);
    }

    // Rate / Margin / Percentage / Share
    if (lowerKey.includes('rate') || lowerKey.includes('margin') || lowerKey.includes('pct') || lowerKey.includes('percent') || lowerKey.includes('growth')) {
      return formatPercentage(num);
    }

    // Count / Quantity / ID
    if (lowerKey.endsWith('_id') || lowerKey === 'id' || lowerKey.includes('year') || lowerKey.includes('code')) {
      return String(value);
    }

    if (Number.isInteger(num)) {
      return num.toLocaleString();
    }

    return num.toLocaleString(undefined, { maximumFractionDigits: 2 });
  }

  return String(value);
}

File: report-generator/app/utils/formatting.py
"""
Formatting utilities for numbers, currencies, dates, and column headers.
Specialized for Enterprise Analytics reports with Indian currency formatting.
"""

from typing import Any, Union
import re
from datetime import datetime


def format_indian_number(value: Union[int, float, str], decimals: int = 2) -> str:
    """
    Format a number using the Indian numbering system (Lakhs, Crores):
    e.g. 2039938 -> "20,39,938.00", 475275 -> "4,75,275.00"
    """
    if value is None or value == "":
        return "-"

    try:
        num = float(value)
    except (ValueError, TypeError):
        return str(value)

    is_negative = num < 0
    num = abs(num)

    # Format with requested decimals
    fmt_str = f"{num:.{decimals}f}"
    parts = fmt_str.split(".")
    integer_part = parts[0]
    decimal_part = f".{parts[1]}" if len(parts) > 1 and decimals > 0 else ""

    # Grouping logic for Indian numbering: last 3 digits, then groups of 2
    if len(integer_part) <= 3:
        formatted_int = integer_part
    else:
        last3 = integer_part[-3:]
        rest = integer_part[:-3]
        groups = []
        while len(rest) > 2:
            groups.append(rest[-2:])
            rest = rest[:-2]
        if rest:
            groups.append(rest)
        groups.reverse()
        formatted_int = ",".join(groups) + "," + last3

    result = f"{formatted_int}{decimal_part}"
    return f"-{result}" if is_negative else result


def format_currency(value: Any, symbol: str = "₹", decimals: int = 2) -> str:
    """Format value as currency using Indian notation."""
    if value is None or value == "":
        return "-"
    formatted = format_indian_number(value, decimals=decimals)
    if formatted.startswith("-"):
        return f"-{symbol}{formatted[1:]}"
    return f"{symbol}{formatted}"


def format_percentage(value: Any, decimals: int = 1) -> str:
    """Format decimal or integer value as percentage."""
    if value is None:
        return "-"
    try:
        num = float(value)
        # If between 0 and 1 (non-zero), assume decimal fraction
        if 0 < abs(num) <= 1.0 and not isinstance(value, int):
            num = num * 100
        return f"{num:.{decimals}f}%"
    except (ValueError, TypeError):
        return str(value)


def humanize_column_name(col_name: str) -> str:
    """Convert snake_case or technical column names to readable titles."""
    if not col_name:
        return ""
    cleaned = col_name.replace("_", " ").strip()
    words = cleaned.split()
    capitalized = []
    acronyms = {"id", "sql", "mcp", "db", "kpi", "hrms", "crm", "erp", "usd", "inr", "pct"}
    for w in words:
        if w.lower() in acronyms:
            capitalized.append(w.upper())
        else:
            capitalized.append(w.capitalize())
    return " ".join(capitalized)


def is_currency_column(col_name: str) -> bool:
    """Detect if column represents financial/currency metrics."""
    lower = col_name.lower()
    keywords = ["revenue", "sales", "amount", "price", "cost", "budget", "spend", "salary", "earnings", "gross", "margin", "total"]
    return any(k in lower for k in keywords) and "count" not in lower and "pct" not in lower and "percentage" not in lower and "rate" not in lower


def is_percentage_column(col_name: str) -> bool:
    """Detect if column represents percentage / share metrics."""
    lower = col_name.lower()
    return "pct" in lower or "percent" in lower or "share" in lower or "ratio" in lower or "rate" in lower


def format_cell_value(val: Any, col_name: str = "", data_type: str = "") -> str:
    """
    Format a table cell value based on column name, data type, and value inspection.
    """
    if val is None or val == "":
        return "-"

    # If already string, check if it's an ISO date
    if isinstance(val, str):
        if re.match(r"^\d{4}-\d{2}-\d{2}", val):
            try:
                dt = datetime.fromisoformat(val.replace("Z", "+00:00"))
                return dt.strftime("%d %b %Y")
            except Exception:
                return val
        return val

    # Boolean
    if isinstance(val, bool):
        return "Yes" if val else "No"

    # Numeric formatting
    if isinstance(val, (int, float)):
        if is_currency_column(col_name):
            return format_currency(val, symbol="₹", decimals=2 if isinstance(val, float) or val % 1 != 0 else 0)
        elif is_percentage_column(col_name):
            return format_percentage(val, decimals=1)
        elif isinstance(val, int):
            return format_indian_number(val, decimals=0)
        else:
            return format_indian_number(val, decimals=2)

    return str(val)
